### PR TITLE
Add rule S1515 (FunctionDefinitionInsideLoop) for JS/TS (new rule for TS)

### LINKED ISSUE #1802: Add rule S1515 (FunctionDefinitionInsideLoop) for JS/TS

### PR DESCRIPTION
**Pull request by quentin-jaquier-sonarsource**
Fixes #1802

### CODE REVIEW

**Comment by quentin-jaquier-sonarsource** [comment]
I inspired myself from both the old Java implementation and Eslint core rule [no-loop-func](https://eslint.org/docs/rules/no-loop-func) implementation.

File: its/ruling/src/test/expected/js/qunit/javascript-S1515.json [hunk 1]
```diff
@@ -1,5 +0,0 @@
-{
-'qunit:src/cli/run.js':[
-59,
```
Comment (quentin-jaquier-sonarsource) [inline its/ruling/src/test/expected/js/qunit/javascript-S1515.json]: The code looks good to me, and is not reported either by [no-loop-func](https://eslint.org/docs/rules/no-loop-func), I think it's fine to not report it.

File: eslint-bridge/tests/rules/function-inside-loop.test.ts [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+          
+          for (var i = 0; i < 10; i++) {
+              funs[i] = (function * () {       // Noncompliant
+                  return function() {           // OK
+                      return i;
+                  };
+              }(i));
+          }
+      `,
+      errors: [
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/function-inside-loop.test.ts]: why  4 `// Noncompliant` in the code and just 2 issues?

File: eslint-bridge/tests/rules/function-inside-loop.test.ts [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    {
+      code: `
+      while (true) {
+        new Promise((resolve) => setTimeout(resolve, 5000))
+        await ping()
+      }
+      `,
+    },
+  ],
+  invalid: [],
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/function-inside-loop.test.ts]: can we add invalid test for TS?

File: eslint-bridge/tests/rules/function-inside-loop.test.ts [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+const ruleTesterTS = new RuleTester({
+  parserOptions: { ecmaVersion: 2018, sourceType: "module" },
+  parser: require.resolve("@typescript-eslint/parser"),
+});
+
+ruleTesterTS.run("Functions should not be defined inside loops", rule, {
+  valid: [
+    {
+      code: `
+      while (true) {
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/function-inside-loop.test.ts]: why this test for TS?

File: eslint-bridge/src/rules/function-inside-loop.ts [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+  "every",
+  "some",
+  "reduce",
+  "reduceRight",
+  "sort",
+];
+
+export const rule: Rule.RuleModule = {
+  create(context: Rule.RuleContext) {
+    const functionAndLoopScopes: estree.Node[] = [];
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/function-inside-loop.ts]: why storing this structure? can't we rely on `parents` of context?

File: eslint-bridge/src/rules/function-inside-loop.ts [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+          const mainLocation = getMainFunctionTokenLocation(
+            node as estree.Function,
+            getParent(context),
+            context,
+          );
+
+          const loopNode = peek(functionAndLoopScopes);
+
+          if (
+            mainLocation &&
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/function-inside-loop.ts]: why this condition? It seems to me that mainLocation is always available. And we can move its computation inside this `if`

File: eslint-bridge/tests/rules/function-inside-loop.test.ts (hunk 4 above)
Comment (quentin-jaquier-sonarsource) [reply eslint-bridge/tests/rules/function-inside-loop.test.ts]: This code can not be parsed by JS parser, but in fact it's due to `await ping()`.
Sine this part of the test code is not critical, I removed it and used only one RuleTester.

File: eslint-bridge/tests/rules/function-inside-loop.test.ts (hunk 3 above)
Comment (quentin-jaquier-sonarsource) [reply eslint-bridge/tests/rules/function-inside-loop.test.ts]: I removed RuleTester for TS, it is not needed anymore.

File: eslint-bridge/src/rules/function-inside-loop.ts (hunk 6 above)
Comment (quentin-jaquier-sonarsource) [reply eslint-bridge/src/rules/function-inside-loop.ts]: Correct, I took this code from another rule, but it seems indeed that mainLocation is always available, I refactored the code.

File: eslint-bridge/src/rules/function-inside-loop.ts (hunk 5 above)
Comment (quentin-jaquier-sonarsource) [reply eslint-bridge/src/rules/function-inside-loop.ts]: Good point

**Comment by quentin-jaquier-sonarsource** [comment]
@vilchik-elena In addition to your comments, I tried to search for enclosing loop outside of a local scope (see last commit), I think the results are good enough, let me know what you think.

**Pull request merged**